### PR TITLE
Fix v18 tablets removed from healthcheck when topo server GetTablet call fails

### PR DESCRIPTION
**Pull request by brendar**
This is the v18 fix for this bug: https://github.com/vitessio/vitess/issues/15632

The code for getting tablets for healthchecks was significantly altered between v18 and v19, so even though the bug is similar, the fix is different.

**Description**
In v18, [`TopologyWatcher.loadTablets()`](https://github.com/Shopify/vitess/blob/v18.0.5-shopify-6/go/vt/discovery/topology_watcher.go#L150) roughly works like this:
* Allocates a `newTablets` map
* Pulls a list of tablet aliases from the topology server
* Iterates through the tablet aliases
  * Spawns a goroutine to call GetTablet to get the tablet record
  * If the GetTablet call succeeds, the tablet is added to `newTablets`
* Waits for all the GetTablet goroutines to finish
* Iterates through `newTablets`:
  * Attempts to find each tablet in `tw.tablets` (the current set of tracked tablets)
  * If it finds an existing tablet, it will update info to match, otherwise it will add that tablet to the healthcheck
* Iterates through `tw.tablets` (the current set of tracked tablets)
  *  Attempts to find each tablet in `newTablets`
  * If it does not find the tablet, it will remove it from the healthcheck

So the bug in v18 is that if a GetTablet call fails, the tablet is not added to `newTablets`, and thus removed from the healthcheck when iterating `tw.tablets`.

**Changes**
* Pulls in the memorytopo changes from the v19+ fix ([link](https://github.com/vitessio/vitess/commit/4caa8d55c6ada0e795c974c5b9dddc75c1ea88d2)). These allow us to force topo errors on specific calls for testing.
* Updates `TopologyWatcher.loadTablets()` to handle GetTablet errors by backfilling `newTablets` with the current value from `tw.tablets`. This is the same approach used in the v19+ fix.
* Pulls in the `TestGetTabletErrorDoesNotRemoveFromHealthcheck` test from the v19+ fix
* Adds `TestGetTabletNoNodeErrorRemovesFromHealthcheck` to cover the case where a tablet is removed from the topo server bewteen the call to `getTablets` and the call to `GetTablet` for that alias. That scenario is handled differently in v19+ so I added explicit handling for it in this fix.

### CODE REVIEW

File: go/vt/discovery/topology_watcher.go [hunk 1]
```diff
@@ -190,6 +190,17 @@ func (tw *TopologyWatcher) loadTablets() {
 			topologyWatcherOperations.Add(topologyWatcherOpGetTablet, 1)
 			<-tw.sem // Done; enable next request to run
 			if err != nil {
+				if !topo.IsErrType(err, topo.NoNode) {
```
Comment (shanth96) [inline go/vt/discovery/topology_watcher.go:193]: Would the error be a `NoNode` even when it's a network partition?

Comment (brendar) [reply go/vt/discovery/topology_watcher.go:193]: No, the `NoNode` error means the path does not exist in zookeeper. We would get a `NoNode` error if the tablet was removed from the topo server between the call to get all the tablet aliases (`tw.getTablets(tw)`) and the call to get that tablet (`tw.topoServer.GetTablet(tw.ctx, alias)`). So we allow the tablet to be removed from the healthcheck on a `NoNode` error.

File: go/vt/topo/memorytopo/memorytopo.go [hunk 2]
```diff
@@ -349,3 +379,24 @@ func (f *Factory) recursiveDelete(n *node) {
 		f.recursiveDelete(parent)
 	}
 }
+
+func (f *Factory) AddOperationError(op Operation, pathPattern string, err error) {
```
Comment (austenLacy) [inline go/vt/topo/memorytopo/memorytopo.go:383]: Where does this get called?

**Review by shanth96 — APPROVED** [review]

File: go/vt/topo/memorytopo/memorytopo.go (hunk 2 above)
Comment (austenLacy) [reply go/vt/topo/memorytopo/memorytopo.go:383]: Oh, never mind. It's in the test. I had that file collapsed while I was reviewing and cmd+f didn't find it.

**Review by austenLacy — APPROVED** [review]

**Pull request merged**